[reland] Make grad point to bucket buffer in DDP to save memory usage

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44344 [reland] Make grad point to bucket buffer in DDP to save memory usage**

reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

zhaojuanmao committed:
[reland] Make grad point to bucket buffer in DDP to save memory usage

reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
+        grad_is_view: when setting to ``True``, grads will be views of all reduce
+                      buckets in reducer, this can reduce peak memory usage, the
+                      saved memory size will be equal to the total grads size.
+                      Also it can avoid coping between grads and the all reduce
```
> Also it can avoid coping between grads and the all reduce buckets in reducer and thus improve training performance.

-> Moreover, it avoids the overhead of copying between gradients and allreduce communication buckets.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
+                      saved memory size will be equal to the total grads size.
+                      Also it can avoid coping between grads and the all reduce
+                      buckets in reducer and thus improve training performance.
+                      When grads are views, detach_() can not be called in your
```
grads -> gradients

double quote `detach_()`

> can not be called in your zero_grad() function, please

cannot be called on the gradients.  If hitting such errors, please

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
+                      buckets in reducer and thus improve training performance.
+                      When grads are views, detach_() can not be called in your
+                      zero_grad() function, please fix it by referring to
+                      the Optimizer.zero_grad() function in torch/optim/optimizer.py
```
double quote all code. 

BTW, can we make `zero_grad()` a link by using `:meth:torch.optim.Optimizer.zero_grad`?

mrshenli reviewed (commented):
Lint errors are real. 

Could you please build the docs to verify the new doc strings are shown correctly? Thanks!

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -316,6 +316,15 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        grad_is_view: when setting to ``True``, grads will be views of all reduce
```
nit: `when set to True` instead of `when setting to True`.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -2749,6 +2748,16 @@ def step_model(model, input, target):
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_accumulate_gradients_module_with_grad_is_view(self):
+        self._test_accumulate_gradients_module(grad_is_view=True)
```
Can we add a test where grad_is_view is True and we're using ddp_comm_hook?

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -1973,13 +1973,15 @@ def tearDown(self):
         return 2
 
-    def _prepare_single_device_module(self, process_group, devices, device_ids, global_batch_size):
+    def _prepare_single_device_module(
```
Shall we turn on more tests? like:

1. `test_multiple_outputs_multiple_backward`
2. `test_accumulate_gradients_no_sync`
3. `test_arbitrary_forward_return_value`
4. etc.

Will they work under this mode?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
+            replica.bucket_views_out[intra_bucket_index];
+        auto& bucket_view_in = replica.bucket_views_in[intra_bucket_index];
+        // If communication_hook is registered, bucket_view_out stores
+        // all reduced results in a newly allocated tensor, copy bucket_view_out
```
all reduced -> allreduced

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
+            if (!grad.defined()) {
+              grad = bucket_view_in;
+            } else {
+              if (!grad.is_alias_of(bucket_view_in)) {
```
do we also need to guard this with `gradient_as_bucket_view_` flag, as we shouldn't print this warning when this mode is off?

zhaojuanmao replied:
when this mode is off, it will go through 'copy_bucket_to_grad' code path, this is only for gradients_as_bucket_view=true

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1015,24 +1104,53 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
+              if (!grad.is_alias_of(bucket_view_in)) {
+                grad.copy_(bucket_view_in);
+                TORCH_WARN_ONCE(
+                    "It is found that it is possible that gradient is not  "
+                    "alias of bucket_view_in here when there are two  "
+                    "different parameter variables in the model share "
+                    "the same storage. For example, param0 and param1 "
+                    "share the same gradient variable grad0. grad0 "
+                    "firstly points to bucket_view_in0 when param0 gradient "
+                    "is ready, then grad0 points to bucket_view_in1 when "
+                    "param1 gradient is ready during  "
+                    "''mark_variable_ready_dense''. In this case, grad0 "
+                    "is not alias of bucket_view_in0 here, although it can "
+                    "still copy bucket_view_in0 to grad0. If you have this "
+                    "case, please check whether it is expected in your "
+                    "application.");
```
Let's reword a bit:

Detected at least one parameter gradient is not the expected DDP bucket view when setting `gradient_as_bucket_view=True`. This can happen when multiple parameters sharing the same gradient. For example, param0 and param1 share the same gradient grad0. In this case, grad0 would first point to `bucket_view_in0` when param0 is ready. Later, when param1 is ready, it will override grad0 to point to `bucket_view_in1`. However, param0 still expects grad0 to point to `bucket_view_in0`, and hence hit this warning. If you saw this message, please double-check if the above situation is expected for your application.

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -230,6 +232,19 @@ class Reducer {
   // with the result of `future_work`.
   void populate_bucket_views_out(BucketReplica& replica, at::Tensor& tensor);
 
+  // If gradient_as_bucket_view_ is false, after all reduce buckets,
```
all reduce -> allreduce

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -230,6 +232,19 @@ class Reducer {
+      bool global_unused);
+  // Check layout of grad and bucket_view before calling copy_grad_to_bucket
+  void check_grad_layout(const at::Tensor& grad, const at::Tensor& bucket_view);
+  // If gradient_as_bucket_view_ is false, before all reduce buckets,
```
ditto

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        gradient_as_bucket_view: this is a prototype feature. When set to ``True``,
```
`gradient_as_bucket_view` -> `gradient_as_bucket_view (bool)`

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      gradients will be views pointing to different offsets of
+                      allreduce communication buckets. This can reduce peak memory
+                      usage, where the saved memory size will be equal to the total
+                      grads size. Moreover, it avoids the overhead of copying
```
grads -> gradients

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      usage, where the saved memory size will be equal to the total
+                      grads size. Moreover, it avoids the overhead of copying
+                      between gradients and allreduce communication buckets.
+                      When gradients are views, "detach_()" cannot be called on the
```
by double quote, I was actually referring to double italic quote, i.e. 
```
``code``
```
This will make our docgen to recognize it as a code.

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      between gradients and allreduce communication buckets.
+                      When gradients are views, "detach_()" cannot be called on the
+                      gradients. If hitting such errors, please fix it by referring to
+                      the :meth:torch.optim.Optimizer.zero_grad function in
```
I think this should be.

```
:meth:`~torch.optim.Optimizer.zero_grad`
```

mrshenli replied:
I might be wrong, does `:meth:torch.optim.Optimizer.zero_grad` also correctly show as a link?

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      When gradients are views, "detach_()" cannot be called on the
+                      gradients. If hitting such errors, please fix it by referring to
+                      the :meth:torch.optim.Optimizer.zero_grad function in
+                      "torch/optim/optimizer.py" as the solution.
```
```
``torch/optim/optimizer.py``
```

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      gradients. If hitting such errors, please fix it by referring to
+                      the :meth:torch.optim.Optimizer.zero_grad function in
+                      "torch/optim/optimizer.py" as the solution.
+                      Warning! It is also found that 'gradient_as_bucket_view = true'
```
warnings are placed in separate blocks. Please search for `.. warning::` in this file.

mrshenli replied:
```
``gradient_as_bucket_view = true``
```

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      the :meth:torch.optim.Optimizer.zero_grad function in
+                      "torch/optim/optimizer.py" as the solution.
+                      Warning! It is also found that 'gradient_as_bucket_view = true'
+                      does not work as expected when apex.amp is used for
```
```
``apex.amp``
```

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      "torch/optim/optimizer.py" as the solution.
+                      Warning! It is also found that 'gradient_as_bucket_view = true'
+                      does not work as expected when apex.amp is used for
+                      mixed precision training. apex.amp maintained stashed gradients
```
```
``apex.amp``
```

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      mixed precision training. apex.amp maintained stashed gradients
+                      that are used for unscaling gradients. These stashed gradients
+                      are pointed to gradients (will be communication buckets when
+                      'gradient_as_bucket_view = true') before starting new iteration.
```
```
``gradient_as_bucket_view = true``
```

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -315,6 +315,28 @@ class DistributedDataParallel(Module):
+                      the unexpectedly muated stached gradients may result in wrong
+                      results. To fix it, these stashed gradients should not be pointed
+                      to gradients, instead they should be copied from gradients when
+                      'gradient_as_bucket_view = true'.
```
```
``gradient_as_bucket_view = true``
```

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

Master was broken and is fixed now. Could you please rebase to the latest commit and wait for all tests to pass before landing? Thanks!

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao commented:
failures are not related

zhaojuanmao committed:
Update on "[reland] Make grad point to bucket buffer in DDP to save memory usage"


reland #41954

Add one argument in DDP API to enable/disable letting grads pointing  to views. When it is disabled, behavior is the same as DDP right now; when it is enabled, Make both variable.grad() and grad in distautograd context point to bucket buffer in DDP to save memory usage.
In this case, grad will be view of bucket buffer tensors, in order to make it compatiable with optimizer.zero_grad(), we
made changes in #41283.

Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

Differential Revision: [D23588186](https://our.internmc.facebook.com/intern/diff/D23588186/)

[ghstack-poisoned]

zhaojuanmao commented:
merge conflict, rebase
